akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -45,28 +104,138 @@ class AttributesSpec extends StreamSpec {
+      val attributes =
+        Source.fromGraph(new AttributesSource(Attributes.name("original-name")))
+          .withAttributes(Attributes.name("replaced")) // this actually replaces now
+          .toMat(Sink.head)(Keep.right).withAttributes(Attributes.name("whole-graph"))
+          .run()
+          .futureValue
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "make the attribues on Flow.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.maybe
+          .viaMat(
+            Flow.fromGraph(new AttributesFlow(Attributes.name("original-name")))
+              .withAttributes(Attributes.name("replaced")) // this actually replaces now
+          )(Keep.right)
+          .withAttributes(Attributes.name("source-flow"))
+          .toMat(Sink.ignore)(Keep.left)
+          .withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "make the attribues on Sink.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.maybe.toMat(
+          Sink.fromGraph(new AttributesSink(Attributes.name("original-name")))
+            .withAttributes(Attributes.name("replaced")) // this actually replaces now
+        )(Keep.right)
+          .withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "use the initial attributes for dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the least specific dispatcher when specified directly around the graph stage" in {
+      val dispatcher =
+        Source.fromGraph(
+          // directly on stage
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher")
+            .addAttributes(ActorAttributes.dispatcher("my-dispatcher")))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the least specific dispatcher when specified further out from the stage" in {
+      val dispatcher =
+        // on the source returned from graph
+        Source.fromGraph(
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher"))
+          .addAttributes(ActorAttributes.dispatcher("my-dispatcher"))
```

Comment:
what happens if that is removed and few flow stages are added here and then `.addAttributes(ActorAttributes.dispatcher("my-dispatcher"))` ?
where are the async boundaries added?

Even if would only be sugar to adding an attribute I think it would be nice if `.async` took an optional dispatcher parameter

Reply:
Yeah. Will play around with more complex flows next. My focus will be to see if I can just get the current scheme to work and leave the reinventing/prototyping new things to @2m and @ktoso

Reply:
Sounds good, I'll explore my idea then and we'll see what's managable/understandable :)